src/darwin/Framework/CHIP/CHIPDevice.mm

```diff
@@ -173,6 +202,898 @@ - (void)subscribeWithQueue:(dispatch_queue_t)queue
+                return CHIP_ERROR_INVALID_ARGUMENT;
+            }
+            NSNumber * elementTag = element[kCHIPTagKey];
+            id elementValue = element[kCHIPValueKey];
+            if (!elementValue) {
+                CHIP_LOG_ERROR("Error: Array element to encode has corrupt value: %@", element);
+                return CHIP_ERROR_INVALID_ARGUMENT;
+            }
+            ReturnErrorOnFailure(EncodeTLVFromObject(elementValue, writer,
+                elementTag ? chip::TLV::ContextTag((uint8_t)[elementTag unsignedCharValue]) : chip::TLV::AnonymousTag()));
+        }
+        ReturnErrorOnFailure(writer.EndContainer(outer));
+        return CHIP_NO_ERROR;
+    }
+    CHIP_LOG_ERROR("Error: Unsupported type to encode: %@", typeName);
+    return CHIP_ERROR_INVALID_ARGUMENT;
+}
+
+// Callback type to pass attribute value as an NSObject
+typedef void (*NSObjectAttributeCallback)(void * context, id value);
+typedef void (*CHIPErrorCallback)(void * context, CHIP_ERROR error);
+
+// Rename to be generic for decode and encode
+class NSObjectData {
+public:
+    NSObjectData()
+        : decodedObj(nil)
+    {
+    }
+    NSObjectData(id obj)
+        : decodedObj(obj)
+    {
+    }
+
+    CHIP_ERROR Decode(chip::TLV::TLVReader & data)
+    {
+        decodedObj = ObjectFromTLV(&data);
+        if (decodedObj == nil) {
+            CHIP_LOG_ERROR("Error: Failed to get value from TLV data for attribute reading response");
+        }
+        return (decodedObj) ? CHIP_NO_ERROR : CHIP_ERROR_DECODE_FAILED;
+    }
+
+    CHIP_ERROR Encode(chip::TLV::TLVWriter & writer, chip::TLV::Tag tag) const
+    {
+        return EncodeTLVFromObject(decodedObj, writer, tag);
+    }
+
+    static constexpr bool kIsFabricScoped = false;
+
+    static bool MustUseTimedInvoke() { return false; }
+
+    id _Nullable GetDecodedObject() const { return decodedObj; }
+
+private:
+    id _Nullable decodedObj;
+};
+
+// Callback bridge for NSObjectAttributeCallback
+class NSObjectAttributeCallbackBridge : public CHIPCallbackBridge<NSObjectAttributeCallback> {
+public:
+    NSObjectAttributeCallbackBridge(
+        dispatch_queue_t queue, CHIPDeviceResponseHandler handler, CHIPActionBlock action, bool keepAlive = false)
+        : CHIPCallbackBridge<NSObjectAttributeCallback>(queue, handler, action, OnSuccessFn, keepAlive) {};
+
+    static void OnSuccessFn(void * context, id value) { DispatchSuccess(context, value); }
+};
+
+// Subscribe bridge for NSObjectAttributeCallback
+class NSObjectAttributeCallbackSubscribeBridge : public NSObjectAttributeCallbackBridge {
+public:
+    NSObjectAttributeCallbackSubscribeBridge(dispatch_queue_t queue,
+        void (^handler)(NSDictionary<NSString *, id> * _Nullable value, NSError * _Nullable error), CHIPActionBlock action,
+        SubscriptionEstablishedHandler establishedHandler)
+        : NSObjectAttributeCallbackBridge(queue, (CHIPDeviceResponseHandler) handler, action, true)
```

Comment:
Because the handler has more specific type for its values, which was preferred during previous code review. CHIPDeviceResponseHandler has more generic type parameter which means the block type is more specific and hence cast was required.